### PR TITLE
[Transformers backend] Enable QKV-Fuser for Gemma4

### PR DESCRIPTION
**Pull request by bohnstingl**
## Purpose

Enable the Transformers-backend QKV fuser to fire on **Gemma 4** attention. The
fuser merges `q(x), k(x), v(x)` into a single `QKVParallelLinear` GEMM by
rewriting the module's forward *source*, but Gemma 4's attention broke three of
its assumptions:

1. **`v_proj` is referenced beyond its single call** — Gemma 4 guards the call
   with a presence check (`v_proj is not None`). The old rewrite required each
   projection to appear exactly once, so it refused.
2. **q/k/v calls are split across branches** — `q` sits at body level while
   `k`/`v` live inside the `else` of `if self.is_kv_shared_layer:`, so no single
   block contains all three calls.
3. **The forward mutates a container** — `shared_kv_states[...] = ...` item
   assignment ended the fx trace early.

Changes:

- **Fold existence guards to a constant.** The fuser only binds instances where
  the projection exists as a truthy `nn.Linear` (guaranteed by the match, the
  cache key, and `validate`), so a guard testing its presence has a constant
  value. `bypass_existence_guard` folds `is` / `is not None` comparisons, in
  either operand order, to the right bool, and folds a **bare truthiness** test
  (`... if self.v_proj else ...`, `not self.v_proj`, `if self.v_proj:`) to
  `True`. `==` / `!=` are deliberately **not** folded: `is_linear` accepts
  `nn.Linear` subclasses that may override `__eq__` / `__ne__`, so equality is
  not guaranteed to track identity. Any other surviving reference (e.g.
  `isinstance(self.v_proj, ...)`, `self.v_proj.weight`, an `and` / `or` operand
  whose value can escape) is not foldable, so the fuser refuses rather than
  change semantics.
- **Hoist the fused GEMM to the dominating block.** `block_chain` replaces
  `innermost_block`, returning the path of `(block, index)` pairs to each call;
  the fused GEMM is inserted before the earliest call, in the innermost block
  common to all three. Live branches survive the rewrite.
- **Trace past item assignment.** `_MetaProxy.__setitem__` records
  `obj[key] = value` as `operator.setitem` so tracing continues.
- **Lift both safety checks onto `StackedFuser`.** Every stacked fuser deletes
  the projections it merges, so guard folding is not specific to QKV:
  `StackedFuser._unguarded_calls` now resolves each projection's call and folds
  any guard on it, and is used by `QKVFuser`, `MergedColumnParallelFuser`,
  `GLUFuser` and `MLAFuser`. Alongside it, `_check_input_stable` replaces
  `MergedColumnParallelFuser`'s structural whitelist with a dataflow check: the
  merged GEMM is evaluated once at the insertion point, so it must read what the
  last call would have. `aliasing_names` tracks the input's names plus anything
  that may alias them (a view shares storage), `written_names` catches rebinding,
  writes through a name, and calls that write an argument (`inplace=`, `out=`,
  `torch.relu_(x)`), and `_returns_fresh` keeps freshly allocated results and
  dead branches from tainting. Both sides over-approximate, since a false hit
  costs a fusion while a miss returns wrong numbers.

The rewrite is a structural transform only — the fused forward is numerically
equivalent to the original (merged GEMM + split with the same weight mapping).

## Limitations

This fuses the layers that own a full q/k/v triple. Two Gemma 4 layer shapes are
intentionally left:

- **Global `k==v` layers** (`attention_k_eq_v` / `use_alternative_attention`):
  `v_proj` is `None` and value reuses key, so only q and k projections exist —
  no triple to merge. Fusing q+k here needs a separate QK fuser (follow-up PR).
- **KV-shared layers** (the last `num_kv_shared_layers`): only `q_proj` exists;
  k/v are read from `shared_kv_states`, so there is nothing to fuse.

Checkpoints using the defaults (`attention_k_eq_v=False`,
`num_kv_shared_layers=0`) have only standard layers, so all of them fuse.

## Test Plan

```bash
.venv/bin/python -m pytest tests/models/transformers/fusers/ -v
```

The whole fusers/ directory, not just test_linear.py: merged_column.py,
glu.py and mla.py all changed.

New/extended coverage in test_linear.py:

- test_fuses_gemma4_qkv_obstacles — guarded v_proj (None, ternary
  truthiness, statement-form if) and branch-split k/v all fuse, with a
  numerics-equivalence check.
- test_qkv_refuses_unsound_rewrites — a rebound input, in-place mutation
  (x.mul_, x[...] = 0), mutation through a view, a functional in-place call
  (F.relu(x, inplace=True)), and a non-guard v_proj reference all fail
  closed.
- test_fuses_real_gemma4_attention — the real Gemma4TextAttention fuses, for
  both the sliding and full-attention layers (which differ in head_dim).
- test_bypass_existence_guard_* — every None-comparison form and operand
  order folds to the correct bool; bare truthiness folds to True; == / !=
  and other non-guard references raise.
- test_merged_column_fuser_folds_existence_guard and
  test_merged_column_fuser_rejects_input_mutation — guard folding and the
  input-stability check are exercised through a non-QKV stacked fuser.

In test_mla.py:

- test_update_forward_rewrites_real_attention — the real
  DeepseekV2Attention source rewrite survives the new check, for both q-LoRA
  configurations.

## Test Result

tests/models/transformers/fusers/            105 passed, 1 skipped
tests/models/transformers/fusers/test_linear.py   58 passed, 1 skipped

End-to-end on the real checkpoint (`google/gemma-4-E2B-it`, CPU):

- Both fusers fire: `q_proj + k_proj + v_proj -> qkv_proj (QKVParallelLinear)`
  and `gate_proj + up_proj -> gate_up_proj (MergedColumnParallelLinear)`.
- Greedy output is **byte-identical** to the unfused native `gemma4.py` path on
  the same prompt, which rules out mis-sharded q/k/v weights rather than only
  proving the rewritten AST is equivalent.
- Chat-templated generation is correct: "What is the capital of France?" answers
  `Paris`.

Run with `limit_mm_per_prompt={'audio': 0}`, needed because
`get_max_audio_tokens` cannot read a max length from `Gemma4AudioConfig`. That
is a pre-existing gap in the Transformers-backend multimodal path, unrelated to
this PR, and I will file it separately.

This is a single greedy prompt on CPU, not an eval sweep. Happy to run lm_eval
on a GPU box if a reviewer wants an accuracy number.

cc @hmellor

### CODE REVIEW

File: vllm/model_executor/models/transformers/fusers/qkv.py [hunk 1]
```diff
@@ -36,6 +36,65 @@
...
+
+    - `self.<name> is/is not/==/!= None`: `is None`/`== None` to `False`,
+      `is not None`/`!= None` to `True`.
+    - a bare truthiness test (`... if self.<name> else ...`, `not self.<name>`):
+      the reference itself to `True`.
+
+    Any other surviving reference escapes the projection's value, which no longer
+    exists after fusion, so refuse rather than change semantics."""
+    for node in ast.walk(funcdef):
+        if not (isinstance(node, ast.Compare) and len(node.ops) == 1):
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/fusers/qkv.py]: The `len(node.ops) == 1)` would skip

```py
if self.<name>:
    something
```

right?

File: vllm/model_executor/models/transformers/fusers/qkv.py [hunk 2]
```diff
@@ -36,6 +36,65 @@
...
+        if (
+            isinstance(node, ast.UnaryOp)
+            and isinstance(node.op, ast.Not)
+            and node.operand is ref
+        ):
+            return True
+    return False
+
+
+def _fold_existence_guard(
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/fusers/qkv.py]:
```suggestion
def _bypass_existence_guard(
```

File: vllm/model_executor/models/transformers/fusers/qkv.py [hunk 3]
```diff
@@ -36,6 +36,65 @@
...
+    """Fold a guard on `self.<name>`'s existence to its constant value.
+
+    The fuser deletes `self.<name>` and binds only instances where it exists as a
+    truthy `nn.Linear` (guaranteed by the match, the cache key, and `validate`),
+    so a guard testing its presence is a fusion invariant. Two forms are folded:
+
+    - `self.<name> is/is not/==/!= None`: `is None`/`== None` to `False`,
+      `is not None`/`!= None` to `True`.
+    - a bare truthiness test (`... if self.<name> else ...`, `not self.<name>`):
+      the reference itself to `True`.
```
Comment (hmellor) [inline vllm/model_executor/models/transformers/fusers/qkv.py]: I would inline these comments with the relevant `ast` code to help readers understand what's happening

Comment (bohnstingl) [reply vllm/model_executor/models/transformers/fusers/qkv.py]: I added inline comments

File: vllm/model_executor/models/transformers/fusers/qkv.py (hunk 1 above)
Comment (bohnstingl) [reply vllm/model_executor/models/transformers/fusers/qkv.py]: Correct, but this was a deliberate choice. I think currently there isn't a network that does a "bare truthiness" check via `_in_boolean_context`, but I incorporated it and it is now also supported. I also added a test to cover it.

File: vllm/model_executor/models/transformers/fusers/qkv.py [hunk 4]
```diff
@@ -126,15 +196,40 @@ def update_forward(self, module: nn.Module) -> None:
...
+            for node in ast.walk(calls[0].args[0])
+            if isinstance(node, ast.Name) and isinstance(node.ctx, ast.Load)
+        }
+        region = block[insert_index : max(indices) + 1]
+        rebound = {
+            node.id
+            for stmt in region
+            for node in ast.walk(stmt)
+            if isinstance(node, ast.Name) and isinstance(node.ctx, (ast.Store, ast.Del))
+        }
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply vllm/model_executor/models/transformers/fusers/qkv.py]: Good catch as well. I integrated an additional check for mutations. The QKV fusing is now rejected when mutations are done on the input to the fused GEMM or when the input is rebound (pre-existing path).

File: vllm/model_executor/models/transformers/fusers/qkv.py [hunk 5]
```diff
@@ -36,6 +36,65 @@
...
+    - a bare truthiness test (`... if self.<name> else ...`, `not self.<name>`):
+      the reference itself to `True`.
+
+    Any other surviving reference escapes the projection's value, which no longer
+    exists after fusion, so refuse rather than change semantics."""
+    for node in ast.walk(funcdef):
+        if not (isinstance(node, ast.Compare) and len(node.ops) == 1):
+            continue
+        (op,), (right,) = node.ops, node.comparators
+        if not isinstance(op, (ast.Is, ast.IsNot, ast.Eq, ast.NotEq)):
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply vllm/model_executor/models/transformers/fusers/qkv.py]: Very sensible catch. I adopted this approach and now only fold `is` and `is not`.

**Review by hmellor — APPROVED** [review]
